Update menu explainer (#1215)
openui/open-ui: site/src/pages/components/menu.explainer.mdx

`<menubar>`

* Menubars commonly represent a series of `<menuitem>`s that are used to open their own submenus (menulists) of commands.

> * Unlike menulists, menubars are rendered **in-page** rather than as a popover, so their menuitems are persistently available.

Reviewer: we could consider saying “part of normal flow” here instead of “in-page”? (like [in CSS](https://developer.mozilla.org/en-US/docs/Learn_web_development/Core/CSS_layout/Introduction#normal_layout_flow))

Author: that's fine by me.  i just kept that term as it was previously written

Reviewer: ```suggestion
* Unlike menulists, menubars are rendered part of normal flow rather than as a popover, so their menuitems are persistently available.
```

* Menubars, and thus their menuitems, are commonly rendered horizontally.
* Supports :disabled, disabled attribute. If a menubar is disabled then all of its descendent menuitems are also disabled.

`<menulist>`

* Menulist should be used to represent a listing, or submenu listing of commands.
* In a menulist, a menuitem's most common purpose is to invoke actions (commands) or other submenus (menulists).
* Menulists are commonly rendered as a **popover**. Submenus are also commonly rendered as adjacent popovers on larger viewports. 
  * On smaller viewports, submenus may overlay or replace the higher-level menulist from which they were invoked. 

>   * A "back" menuitem may be rendered to allow a user to "navigate up" to the higher-level menulist.

Reviewer 2: Great idea. I have seen such use cases especially for mobile web app, where there isn't enough space to show multiple nested submenus (or to have better UX).

Author: wow... completely dropped the ball on noticing these questions...

this was added because i think it'd be great if UAs provided this sort of styling - or at least made it very easy to provide this sort of styling by default. 

styling for menu popups for small viewports / zoomed in viewports is often overlooked, which can make it rather difficult if not impossible for low vision users to interact with the menuitems when authors haven't been diligent at making sure their menus actually appear in the viewport.

but, i realize this might be more effort than all browsers want to take on.  so I said "may" to provide wiggle room.  maybe mobile browsers do this and desktop browsers leave it up to authors to remember.   shrug

* The menulist popover is anchored to the button or menuitem which invoked it
* Supports :disabled, disabled attribute. If a menulist is disabled then all of its descendent menuitems are also disabled.

**Question:** are the `open` or `name` attributes needed for `menubar` or `menulist`?

> ???

Reviewer 2: The `open` attribute is useful if we want to default the menulist as open. But maybe we should avoid that, and be explicitly to only open with user activation behaviors. It is definitely not necessary on `menubar`.

The `name` attribute is meant to group radio menuitem elements together. It is probably not necessary for `menubar` and `menulist`.

Reviewer 3: I agree that `name` isnt' needed if we're doing the `<fieldset>` approach, which is our current plan. I don't see a need for `open` attribute unless people want to be able to set a menu's open status with JavaScript besides just activating it. I guess I'd prefer to wait for such asks to come on, and stick with the simpler, smaller proposal of not including this attribute initially.

Reviewer 4: One thing to be clear on, we cannot do open as a two way binding. Aka it cannot reflect state like details and dialog. So if we want an initiallyopen we should plan to name it accordingly.

Reviewer 4: > Why not?

Because it's a flustercuck for details and especially dialog, this will have many of the issues dialog has. I believe Mason also shares this opinion.

Reviewer 5: I share this opinion. The `open` attribute should be considered a bad idea. The issues stem from the fact that modifying DOM (to add or subtract an attribute) is prohibited or just bad in several places, such as removing an element from the DOM. The best pattern is:

1. A pseudo class like `:open` to reflect state
2. (If needed) an attribute that controls the *initial* state, such as `initiallyopen`.

<img src="/images/menubar-search-input.png" alt="examples of text fields as the first child of menulist popups. The first example shows a text field where a percentage can be adjusted, it is followed by a number of menuitems to zoom or adjust an editable canvas. The second example shows a search text field as the first item in a menulist invoked from a Help menuitem of a menubar. Two additional menuitems follow the search text field in this menulist popup." />

## Questions

### When to use Select vs Menulist

Just like a select element, the menulist element will represent a series of choices to pick from. And with customizable select now available, users might be confused on which one to use.

The `<select>` element is a form-associated element and contains `<option>` elements. Each of these `<option>` elements must have a server-readable value. As a user chooses an option, specific events are triggered (change, input) and a new value is set.

> On the other hand, the proposed `<menulist>` element should be used to display a list of commands/actions. Each `<menuitem>` should be thought of as a command that is activable. That action could be to open a new nested menulist element or to trigger a specific command. It is not generally expected for the invoking element's value or visible label to be modified when activating a `<menuitem>`.  It must adhere to ARIA guidelines for [menu and menubar patterns](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/).

Reviewer: not part of your change, but I don't think “activable” is a word? And it's not very clear what “command that is activable” is. Maybe we can fix at the same time

```suggestion
On the other hand, the proposed `<menulist>` element should be used to display a list of commands/actions. Each `<menuitem>` should be thought of as performing a specific action. That action could be to open a new nested menulist element or to trigger a specific command. It is not generally expected for the invoking element's value or visible label to be modified when activating a `<menuitem>`.  It must adhere to ARIA guidelines for [menu and menubar patterns](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/).
```

> On the other hand, the proposed `<menulist>` element should be used to display a list of commands/actions. Each `<menuitem>` should be thought of as a command that is activable. That action could be to open a new nested menulist element or to trigger a specific command. It is not generally expected for the invoking element's value or visible label to be modified when activating a `<menuitem>`.  It must adhere to ARIA guidelines for [menu and menubar patterns](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/).

Reviewer 2: nit: I wrote ARIA guidelines originally, but that is wrong. Should be changed to APG recommendations.

The [`<menu>`](https://html.spec.whatwg.org/multipage/grouping-content.html#the-menu-element) element already exists, but it is very barebones. It represents a "toolbar" but is exposed no differently than an unordered list containing list items, each list item then representing a command (represented by a button element) that the user can perform or activate. There is a lot of nuance between the various menu patterns (menubar, navigation menu, toolbar) that this one element cannot differentiate. It has no special keyboard behavior nor unique ARIA roles mapping beyond the implicit roles of the elements it represents. Due to these legacy restrictions, we recommend not modifying it and instead depend on the new elements `<menubar>` and `<menulist>`.

> You can read more about what the `<menu>` element is and isn’t, [here](https://www.scottohara.me/blog/2021/10/21/menu.html).

Reviewer: maybe we could improve link text?

```suggestion
You can read more about what the `<menu>` element is and isn’t, in Scott O'Hara's blog post [Semantic `<menu>` context](https://www.scottohara.me/blog/2021/10/21/menu.html).
```
